Tasks for lessons 1, 2, 3, 4

govorushkinanton opened the pull request:
*Also added invalid sorting order handling

potyzhnyi committed:
Implemented 'sortStrings(array:string[], sortingOrder:string)' function. Also added invalid sorting order handling

potyzhnyi committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20200821-3_govorushkinanton

jsru commented:
Пожалуйста, не меняйте «системные» файлы.
          Задача не может считаться решенной, если были изменены файлы тестов или другие «системные» файлы.
          Следующие файлы были изменены:
```diff
- 01-intro/1-sum/index.test.js — изменен
```

govorushkinanton commented:
> Пожалуйста, не меняйте «системные» файлы.
> Задача не может считаться решенной, если были изменены файлы тестов или другие «системные» файлы.
> Следующие файлы были изменены:
> 
> ```diff
> - 01-intro/1-sum/index.test.js — изменен
> ```

Без этих изменений этот тест у меня "падал" c ошибкой:
` TypeError: (0 , _index.default) is not a function`

jsru commented:
Пожалуйста, не меняйте «системные» файлы.
          Задача не может считаться решенной, если были изменены файлы тестов или другие «системные» файлы.
          Следующие файлы были изменены:
```diff
- 01-intro/1-sum/index.test.js — изменен
```

potyzhnyi committed:
Implemented 'pick(obj, ...fields function)' function

potyzhnyi committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20200821-3_govorushkinanton

jsru commented:
Пожалуйста, не меняйте «системные» файлы.
          Задача не может считаться решенной, если были изменены файлы тестов или другие «системные» файлы.
          Следующие файлы были изменены:
```diff
- 01-intro/1-sum/index.test.js — изменен
```

potyzhnyi committed:
Implemented 'omit(obj, ...fields function)' function

Added explanations why Object.entries() was not used for the task completion

jsru commented:
Пожалуйста, не меняйте «системные» файлы.
          Задача не может считаться решенной, если были изменены файлы тестов или другие «системные» файлы.
          Следующие файлы были изменены:
```diff
- 01-intro/1-sum/index.test.js — изменен
```

potyzhnyi committed:
Fixed function import

jsru commented:
Поздравляю, все задачи решены абсолютно верно
```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings
+ should correctly sort language-specific characters
+ should correctly sort strings started from uppercase
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```

Добавляю преподавателя (@dosandk) для код-ревью, посмотрим что он скажет :)

dosandk commented on 02-javascript-data-types/1-sort-strings/index.html:
```diff
@@ -0,0 +1,31 @@
+<!DOCTYPE html>
```
в PR не должно быть посторонних файлов - нужно убрать из PR

govorushkinanton replied:
Удалил все html из проекта

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,22 @@
  */
 export function sortStrings(arr, param = 'asc') {
-
+  let sortedArray = [];
```
данная переменная должна быть объявлена через `const`

govorushkinanton replied:
Fixed

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,22 @@
-
+  let sortedArray = [];
+  if (param === 'asc' || param === undefined){
+    Object.assign(sortedArray, arr.sort(function (a, b) {
```
ветки `if` и `else if` дублируют друг друга, за исключением порядка аргументов в `localCompare` - это дублирование необходимо устранить

govorushkinanton replied:
Вынес логику сортировки в отдельный метод `sortArray(array)`

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -4,6 +4,16 @@
+ * from https://learn.javascript.ru/object
+ */
+export const pick = (obj, ...fields) => {
+  let sortedObject = {};
```
переменная должна быть объявлена через `const`

govorushkinanton replied:
Согласен. Fixed

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -4,6 +4,16 @@
+ */
+export const pick = (obj, ...fields) => {
+  let sortedObject = {};
+  fields.forEach(entry=>{
```
Подумайте, возможно следует итерировать `obj`, а не `fields`?

Код, над которым следует подумать:
```
const pick = (obj, ...fields) => {
    const result = {};
    fields.forEach(field => {
        result[`${field}`] = obj[field];
    });
    return result;
};

const obj = {
	foo: 'bar'
};

console.log(pick(obj, 'toString'));
```

govorushkinanton replied:
Согласен. Fixed

dosandk commented on 02-javascript-data-types/3-omit/index.js:
```diff
@@ -4,6 +4,15 @@
+ * from https://learn.javascript.ru/object
+ */
+export const omit = (obj, ...fields) => {
+  fields.forEach(entry=>{
```
аналогичные замечания как в `pick`

govorushkinanton replied:
Fixed

dosandk reviewed (changes requested):
необходимо внести правки согласно комментариям

potyzhnyi committed:
Made changes according to comments

Not sure about duplications - created separate method and reusing it across if/else if

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

potyzhnyi committed:
Made changes according to comments

Not sure about duplications - created separate method and reusing it across if/else if

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

dosandk commented on 01-intro/1-sum/index.js:
```diff
@@ -5,5 +5,6 @@
  */
 export default function sum(m, n) {
-  return n + m;
+  return m + n;
```
good 👍

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,22 @@
  */
 export function sortStrings(arr, param = 'asc') {
+  const sortedArray = [];
+  if (param === 'asc' || param === undefined){
```
В каком случае в данной функции `param` может быть `undefined`?

govorushkinanton replied:
Если аргумент не был передан в функцию. Переписал функцию

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,22 @@
 export function sortStrings(arr, param = 'asc') {
+  const sortedArray = [];
+  if (param === 'asc' || param === undefined){
+    Object.assign(sortedArray, sortArray(arr, 1));
```
Не стоит использовать `Object.assign` для копирования массивов, есть более короткие и семантически верные пути сделать это:

* [...arr]
* arr.slice()

govorushkinanton replied:
Переписал функцию. Сделал копию с `concat()`

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,22 @@
+    Object.assign(sortedArray, sortArray(arr, -1));
+    return sortedArray;
+  }
+  else {
```
лучше хранить условия в виде объекта, вот так:

```
const directions = {
  asc: 1,
  desc: -1
};
```

Тогда, в случае, если действительно необходимо бросить ошибку и предупредить пользователя о том что такого параметра сортировки нет, это будет гораздо проще сделать:

```
const direction = directions[param];

if (!direction) {
  throw new Error(`Invalid sorting order specified. Please use either ${Object.keys(directions).join(' ')}`)
}

  return sortArray([...arr], direction);
}
```

govorushkinanton replied:
Переписал 👍

dosandk replied:
Не вижу изменений. Вот от этой конструкции `Object.assign(sortedArray, sortArray(arr, -1));` следует избавиться, нам не нужен `Object.assign` для копирования массива

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -4,6 +4,17 @@
+ */
+export const pick = (obj, ...fields) => {
+  const sortedObject = {};
+  for(let param in obj){
```
посмотрите пожалуйста почему не стоит использовать `for...in` для обхода объектов https://jstips.space/loops/

govorushkinanton replied:
Прочитал, понял, переписал :) For in может использоваться с проверкой на `hasOwnProperty` вместо for...of? Мне лично так легче читается

dosandk commented on 02-javascript-data-types/3-omit/index.js:
```diff
@@ -4,6 +4,17 @@
+ */
+export const omit = (obj, ...fields) => {
+  const sortedObject = {};
+  for (let param in obj){
```
посмотрите пожалуйста почему не стоит использовать `for...in` для обхода объектов https://jstips.space/loops/

govorushkinanton replied:
✔️

dosandk reviewed (changes requested):
необходимо внести правки согласно комментариям

potyzhnyi committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20200821-3_govorushkinanton

potyzhnyi committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20200821-3_govorushkinanton

potyzhnyi committed:
- Made changes according to comments for tasks of lesson 2;
 - Task 'invert object' lesson 3;
 - Task 'Chart' of lesson 4

potyzhnyi committed:
- Made changes according to comments for tasks of lesson 2;
 - Task 'invert object' lesson 3;
 - Task 'Chart' of lesson 4

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

potyzhnyi committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20200821-3_govorushkinanton

potyzhnyi committed:
Delete DefaultComponent.js

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
ColumnChart компонент
+ should be rendered correctly
+ should have ability to define label
+ should have ability to define link
+ should have ability to define total value
+ should have property 'chartHeight
+ should render data correctly
+ should have ability to be updated by new "data" values (should re-render only body with charts columns)
+ should have loading indication if data wasn't passed 
+ should have ability to be removed
```
```diff
createGetter
- should return existed properties
- should return "undefined" for properties which does not exist
- gets obj.property
- returns undefined for no property
- gets obj.nested.property
- returns undefined for no nested.property
- gets more.nested.property
```
```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings
+ should correctly sort language-specific characters
+ should correctly sort strings started from uppercase
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
invertObj
+ should swap keys and values and return new object
+ should return empty object if was passed object without values
+ should return "undefined" if object wasn't passed
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```
```diff
uniq
- should return array with uniq values
- should return empty array if was passed array without values
- should return empty array if arguments wasn't passed
```

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,67 @@
+  render() {
+    const chart = document.createElement('div');
+
+    chart.innerHTML = `<div class="column-chart column-chart_loading" style="--chart-height: ${this.chartHeight}">
```
тимплейт лучше хранить в отдельном методе, тогда не будет смешения логики рендера и "простыни" тимплейта

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,67 @@
+
+    this.element = chart.firstElementChild;
+
+    if (this.data.length != 0) {
```
`!=` такого оператора в JS нет, точно также как и такого `==`

govorushkinanton replied:
В книге написано, что "==" есть (поисковой запрос - "оператор равенства") - https://learn.javascript.ru/object

dosandk commented on 04-oop-basic-intro-to-dom/1-column-chart/index.js:
```diff
@@ -1,3 +1,67 @@
+
+    this.element = chart.firstElementChild;
+
+    if (this.data.length != 0) {
```
такие проверки следует сокращать:
```
if (this.data.length) 
```

dosandk commented on 03-objects-arrays-intro-to-testing/2-invert-object/index.js:
```diff
@@ -4,5 +4,13 @@
  * @returns {object | undefined} - returns new object or undefined if nothing did't pass
  */
 export function invertObj(obj) {
```
good 👍

dosandk commented on 02-javascript-data-types/3-omit/index.js:
```diff
@@ -4,6 +4,19 @@
+
+  for (let [key, value] of Object.entries(obj)){
+    if (fields.find(fieldToOmit => fieldToOmit === key)){
+      delete obj[key]
```
не следует удалять значения из объекта! Нужно переписать функцию без использования оператора `delete`

dosandk commented on 02-javascript-data-types/3-omit/index.js:
```diff
@@ -4,6 +4,19 @@
 export const omit = (obj, ...fields) => {
+  const sortedObject = {};
+
+  for (let [key, value] of Object.entries(obj)){
```
[key, value] должны быть объявлены через `const`

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -4,6 +4,19 @@
 export const pick = (obj, ...fields) => {
+  const sortedObject = {};
+
+  for(let [key, value] of Object.entries(obj)){
```
[key, value] должны быть объявлены через const

dosandk reviewed (changes requested):
необходимо внести правки согласно комментариям

potyzhnyi committed:
Made changes according to comments. Added 'Getter' task

potyzhnyi committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20200821-3_govorushkinanton

potyzhnyi committed:
Added solution for 'trimString' task

potyzhnyi committed:
Merge remote-tracking branch 'origin/master'

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

dosandk reviewed (approved):
good 👍
